feature/permissions/src/main/java/com/google/jetpackcamera/permissions/PermissionsViewModel.kt

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.jetpackcamera.permissions
+
+import android.Manifest
+import androidx.compose.runtime.mutableStateListOf
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.google.accompanist.permissions.ExperimentalPermissionsApi
+import com.google.accompanist.permissions.MultiplePermissionsState
+import com.google.accompanist.permissions.isGranted
+import com.google.accompanist.permissions.shouldShowRationale
+import dagger.assisted.Assisted
+import dagger.assisted.AssistedFactory
+import dagger.assisted.AssistedInject
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.launch
+
+/**
+ * A [ViewModel] for [PermissionsScreen]]
+ */
+@OptIn(ExperimentalPermissionsApi::class)
+@HiltViewModel(assistedFactory = PermissionsViewModel.Factory::class)
+class PermissionsViewModel @AssistedInject constructor(
+    @Assisted permissionStates: MultiplePermissionsState
+) : ViewModel() {
+    @AssistedFactory
+    interface Factory {
+        fun create(runtimeArg: MultiplePermissionsState): PermissionsViewModel
+    }
+    private val permissionQueue = mutableStateListOf<PermissionEnum>()
+
+    init {
+        permissionQueue.addAll(getRequestablePermissions(permissionStates))
+    }
+
+    private val _permissionsUiState: MutableStateFlow<PermissionsUiState> =
+        MutableStateFlow(getCurrentPermission())
+    val permissionsUiState: StateFlow<PermissionsUiState> = _permissionsUiState.asStateFlow()
+
+    private fun getCurrentPermission(): PermissionsUiState {
+        if (permissionQueue.isEmpty()) {
+            return PermissionsUiState.AllPermissionsGranted
+        } else {
+            return PermissionsUiState.PermissionsNeeded(permissionQueue.first())
+        }
+    }
+
+    fun dismissPermission() {
+        if (!permissionQueue.isEmpty()) {
+            permissionQueue.removeFirst()
+        }
+
+        viewModelScope.launch {
+            _permissionsUiState.emit(getCurrentPermission())
+        }
```

Comment:
Since `_permissionsUiState` is a `StateFlow`, you can skip the `viewModelScope.launch` and just do `_permissionsUiState.update { }`.